core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryEvaluationContext.java

```diff
@@ -30,7 +30,11 @@
  */
 public interface QueryEvaluationContext {
 
-	public class Minimal implements QueryEvaluationContext {
+	class Minimal implements QueryEvaluationContext {
+
+		private Literal now;
```

Comment:
Has to be volatile. If used in double checked locking. Did this show up in a profile? as it is once per query execution?

Reply:
Yeah, that's right. Should be volatile. I think I was actually considering removing the synchronization all together and removed the use of volatile first and then forgot to remove the synchronization. 

It shows up for the ShaclSail because it generates a lot of queries and each query just returns a handful of results. It's part of how the ShaclSail can analyze the changes in the transaction and only validate the data that is affected.

The main issue is actually that creating a new date is really expensive because we use the GregorianCalendar and XMLGregorianCalendar. Don't really know how to get away from this, but using lazy loading in this class here kinda solves it temporarily for the ShaclSail.

Reply:
Maybe just do something like this.
```java
private final long nowInMillis = System.currentTimeInMillis();
private Literal now;

public Literal getNow() {
    if (now != null) {
         now = SimpleValueFactory.getInstance().createLiteral(new Date(now));
    }
    return now;
}
```
No need to lock as two invocations will give equivalent Literals. We only store a long in every case and no object is created until required.

Reply:
Indeed the one on develop though is not thread safe, the one suggested here is. Let's fix that on a different issue.